### PR TITLE
"Added notebooks that are colab accessible, they should both be runnable from within colab. "

### PR DESCRIPTION
**Pull request by John-Peters-UW**
To run,

- Open colab
- paste in the link to the notebooks (the github links should work, I think) 
- Press run all, or modify the inputs and run the notebooks to play with them

### CODE REVIEW

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:338]:
```suggestion
        "The PDB file is something that we will reference from within the colab through its file path. If you don't need a PDB file for your metl model, simply set this variable to None."
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:379]: Do links not work in this Markdown syntax?

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:406]:
```suggestion
        "Lastly, we will then collect some variants. The code in this notebook supports variants in JSON list format. Either enter a JSON-formatted list of variants in the text box below or upload a variant JSON file in the following cell. Run the cell to active the text entry or file upload."
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:501]:
```suggestion
        "If you would rather upload a file, run the cell below and use it to upload a file. If a file is uploaded, the input above will not be looked at for variants\n"
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:583]:
```suggestion
        "# @title Variant selecting logic (always run this)\n",
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:603]: We should try to be consistent with either metl or METL in the instructions.

```suggestion
        "For biologists, one-based indexing is commonly used. However, METL models were designed to used zero-based indexing. If one-based indexing is needed, select it in the dropdown below."
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:640]:
```suggestion
        "# @title METL predicting loop\n",
```

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [inline notebooks/colab_metl_predicting.ipynb:692]: Can we also display the predictions?

**Review by agitter — CHANGES_REQUESTED** [review]
The notebooks are great and overall worked well for me. I'd like to get some feedback from others who are less familiar with how METL works.

We will need to add these to the notebooks `README.md` and explain how to run them in Colab. We can provide a direct link to open these notebooks in Colab. ColabFold uses `https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb` for example. We should document that the T4 GPU runtime is sufficient to run both notebooks. Is it necessary to connect the runtime in Colab before starting?

Will we keep the original `finetuning.ipynb` as well? We probably can as long as we explain their different purposes.

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb:692]: Yea, I was just worried about the metl models that display embeddings taking up too much screen space. But I can display a cut off view if its going to take too many lines. Even with the scrollable limits it's a lot most of the time.

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:484]: If you want to run full-length fine-tuning you have to change it, but it's not really described. We can delete it or add a GUI interface for all of these variables so it's a bit more clear.

In the colab one we could probably delete that line about dealing with apple silicon though, since it's just going to run with CUDA

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:51]: I'll also do some benchmarking here to see if there's _anything_ that can be done to speed it up, but I think right now it's actually fairly consistent with what even a local setup is taking with all of the downloads. 

It's not even the solvers fault (I tried the mamaba one to no effect), it's just a lot of bandwidth being used it seems (Looking at the colab storage allocations it seems like it's many gigabytes, and more than a few. I can get actual numbers though)

**Comment by John-Peters-UW** [comment]
Good feedback :) 

> We will need to add these to the notebooks README.md and explain how to run them in Colab. We can provide a direct link to open these notebooks in Colab. ColabFold uses https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb for example. We should document that the T4 GPU runtime is sufficient to run both notebooks. Is it necessary to connect the runtime in Colab before starting?

From my understanding I think you have to connect to it manually if you want to use the GPU. I think it has something to do with them not trying to automatically start billing you if you are allocating like A100's or something

**Comment by John-Peters-UW** [comment]
> Will we keep the original finetuning.ipynb as well? We probably can as long as we explain their different purposes.

Yea. I think we should keep them separate. If you are running the normal one, you don't need the notebook cloning metl again (it's assumed it should be cloned already) and force-downloading all of the dependencies every time you try to run it

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:218]: I think not clogging up the output is sorta polite, so I like the comments. For all the notebooks I've been writing (especially the GUI ones), I have been assuming run all needs to work so I've been putting safe-guards in.

I think assuming run all would prioritize good programming practice so I think it's valuable to always assume it'll be used like that.

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb:979]: Made it so that the output height was limited and it works nicely

File: notebooks/colab_finetuning.ipynb
Comment (agitter) [inline notebooks/colab_finetuning.ipynb:180]: This global link and some others are broken. Some are still relative. Please double check them all.

File: notebooks/colab_finetuning.ipynb
Comment (agitter) [reply notebooks/colab_finetuning.ipynb:635]: Still have this question, but I'm not sure how to control this.

File: notebooks/colab_finetuning.ipynb
Comment (agitter) [reply notebooks/colab_finetuning.ipynb:740]: Missed this suggestion

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [reply notebooks/colab_metl_predicting.ipynb:1]: What do you think about renaming the file?

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [reply notebooks/colab_metl_predicting.ipynb]: Some of these previous suggesions may not have been applied

File: notebooks/colab_metl_predicting.ipynb
Comment (agitter) [reply notebooks/colab_metl_predicting.ipynb:979]: This looks great

**Comment by agitter** [comment]
I made another pass through the notebooks. Some of the orginal comments are still unresolved, including some small typos corrections. I'd like to add these to the readme table before merging.

@samgelman do you want to test these in Colab before we merge?

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:180]: @agitter The broken links were links to python or ipynb files that are broken in the original notebooks as well.

Additionally, the broken global links are actually linking to pages that no longer exist. The only way to get to those files are by going to older commits, when they did exist, like the other notebook does. 

I do not know what you want to do with that data. I fixed the ipynb links, but I will wait on the broken globals because those files are no longer in the current version of the main repository. 

An example of that is this link here for the dms_data that is in the notebook:  https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/data/dms_data

But the actual file was deleted though can be found in an earlier commit:  https://github.com/gitter-lab/metl/tree/45237bbc4c6cfae1d4657930bdb732ec41a3fc79/data/dms_data

Are we supported this still? Does this need to be referenced or talk about? And if it does, why is it not included in the current version of main? 

I'm confused on these so I am leaving them, again, I fixed the broken links to python files and notebooks (that were also broken in the normal finetuning notebook)

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb:1]: I will rename it. That is a better name I think

**Comment by John-Peters-UW** [comment]
> I made another pass through the notebooks. Some of the orginal comments are still unresolved, including some small typos corrections. I'd like to add these to the readme table before merging.
> 
> @samgelman do you want to test these in Colab before we merge?

Were the typo's the links or were there other things that you saw? I can't seem to find anything else. I thought I got all of the typo corrections you put out on the previous round of comments

File: notebooks/colab_finetuning.ipynb
Comment (agitter) [reply notebooks/colab_finetuning.ipynb:180]: Got it. This is a question for @samgelman.

**Comment by agitter** [comment]
> Were the typo's the links or were there other things that you saw? I can't seem to find anything else.

I think a lot of my small comments on the pretraining notebook were masked by GitHub as outdated but weren't incorporated. Try searching "To zero based helper functoin" as one example.

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:635]: Yea. In colab if you expand it and then save it, it will start off expanded I think. I will re-save it and see if it fixes it.

File: notebooks/colab_finetuning.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_finetuning.ipynb:855]: Yes, whenever you are ready to do so we may

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: That will work. I like that

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: Yea, this is temp

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: https://github.com/gitter-lab/metl-pretrained/blob/huggingface_support/huggingface/print_colab_dropdown.py

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: I don't think it is right now. I'll leave this open for now so that we can discuss it later more easily

Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: They aren't all files though. Only 1 is a file, the rest of strings

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: I deleted this. I made it so the user doesn't have to think of that but forgot to get rid of the previous instructions before the dropdown was available

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: When you put the brackets the right way round they do 😂

File: notebooks/colab_metl_predicting.ipynb
Comment (John-Peters-UW) [reply notebooks/colab_metl_predicting.ipynb]: fixed

**Comment by John-Peters-UW** [comment]
> > Were the typo's the links or were there other things that you saw? I can't seem to find anything else.
> 
> I think a lot of my small comments on the pretraining notebook were masked by GitHub as outdated but weren't incorporated. Try searching "To zero based helper functoin" as one example.

Should've gotten them now. There are a few comments that you can answer though in the unresolved comments above

File: notebooks/colab_finetuning.ipynb
Comment (samgelman) [reply notebooks/colab_finetuning.ipynb:180]: In this case, the link https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/data/dms_data is incorrect, it should just be https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp.

**Comment by samgelman** [comment]
I've reviewed both notebooks, and they are working great! 

I have a few minor comments, for the prediction notebook:
- For the "Variant text input" cell, specify whether the example variants are 0-based or 1-based indexing
- For the "Variant file upload" cell, how should the JSON file be formatted? It might be simpler for biologists to be able to provide a list of variants in a text file, one on each line, rather than formatting a JSON file.
- For the "METL predicting loop" cell, the output predictions are labeled as logits, but I'm not sure if that is the correct way to describe them. I usually see logits in the context of classifier networks, and the term refers to the outputs from the final layer of the network, before the softmax / activation. I think I've also seen the term to describe intermediate network layer outputs. For us, I think "predictions" or simply "outputs" would be more accurate.

For the finetuning notebook, there are some broken links. It looks like there may have been a find and replace, but the replacement text erroneously added "data/dms_data/avgfp" to the path.
- https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/data/dms_data should be https://github.com/gitter-lab/metl/tree/main/data/dms_data
- https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/pretrained_models should be https://github.com/gitter-lab/metl/tree/main/pretrained_models
- https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/data/dms_data/avgfp should be https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp
- https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/code/models.py should be https://github.com/gitter-lab/metl/tree/main/code/models.py
- https://github.com/gitter-lab/metl/tree/main/data/dms_data/avgfp/args/pretrain_avgfp_local.txt should be https://github.com/gitter-lab/metl/blob/main/args/pretrain_avgfp_local.txt
and so on.

I suggest fixing the broken links, and then other changes can be discussed / added later in a separate PR (especially the JSON vs. text file upload, I don't think that's a critical issue that should block the merge).

Great job!

**Comment by agitter** [comment]
I reran both notebooks and made a few small changes in 9912989380ebe1246a2e35a92488e424d7ae571b. They both look excellent.

**Pull request merged**
